torch/csrc/jit/codegen/cuda/lower_index_hoist.cpp

```diff
@@ -12,419 +12,275 @@ namespace cuda {
-      continue;
+    if (inputs.back() != input) {
+      changed = true;
     }
+  }
 
-    return false;
+  // If any of the inputs is replaced, then create a new expression whose inputs
+  // are replaced with hoisted input
+  if (changed) {
+    index = IrBuilder::newScalar(*index->getDataType());
+    TORCH_INTERNAL_ASSERT(def->outputs().size() == 1);
+    auto create_fn = def->newObjectFunc();
+    create_fn(index->container(), inputs, {index}, def->attributes());
   }
 
-  return true;
+  if (my_level < parent_level) {
+    insertIndexToLoop(index, loops, my_level);
+  }
+  return index;
 }
 
-std::string CommonIndexKey::toString() const {
-  TORCH_INTERNAL_ASSERT(concrete_indexed_id_ != nullptr);
-  std::stringstream ss;
-  ss << "CommonIndexKey: " << concrete_indexed_id_->toString();
-  ss << ", { ";
-  for (auto loop : used_loops_) {
-    ss << loop->iter_domain()->toString() << " ";
+Val* CommonIndexMap::hoistIndex(
+    Val* index,
+    const std::vector<kir::ForLoop*>& loops) {
+  if (isOptionDisabled(DisableOption::IndexHoist)) {
+    return index;
   }
-  ss << "}";
-  ss << ", { ";
-  for (auto val : loop_index_vals_) {
-    ss << val->toString() << " ";
+  auto hoisted_index = hoistIndex(index, loops, loops.size() + 1);
+  if (hoisted_index == nullptr) {
+    // If index hoisting failed, just return the original index
+    return index;
   }
-  ss << "}";
-  return ss.str();
+  return hoisted_index;
 }
 
-std::pair<Val*, bool> CommonIndexMap::insert(
-    IterDomain* indexed_consumer_id,
-    TensorDomain* consumer_td,
-    TensorDomain* ref_td,
-    const std::unordered_map<IterDomain*, Val*>& ref_index_map,
+void CommonIndexMap::insertIndexToLoop(
+    Val* index,
     const std::vector<kir::ForLoop*>& loops,
-    Val* index) {
-  if (index->definition() == nullptr) {
-    // Only defined val is eligible to hoist
-    return {index, false};
+    int64_t level) {
+  kir::ForLoop* loop = nullptr;
+  if (level > 0) {
+    loop = loops[level - 1];
   }
-
-  const CommonIndexKey key(
-      indexed_consumer_id, consumer_td, ref_td, ref_index_map, loops);
-
-  return tryInsertNewIndex(key, index);
+  common_index_map_[loop].emplace_back(index);
 }
 
-std::pair<Val*, bool> CommonIndexMap::insert(
-    IterDomain* indexed_consumer_id,
-    TensorDomain* consumer_td,
-    const std::vector<IterDomain*>& loop_domains,
-    const std::unordered_map<IterDomain*, Val*>& loop_index_map,
+Val* CommonIndexMap::eliminateCommonSubexpression(
+    Val* index,
     const std::vector<kir::ForLoop*>& loops,
-    Val* index) {
-  if (index->definition() == nullptr) {
-    // Only defined val is eligible to hoist
-    return {index, false};
+    int64_t level) {
+  // For all outer loops, find if that loop already contain `index`.
+  for (auto current_level : c10::irange(level + 1)) {
+    kir::ForLoop* loop = nullptr;
+    if (current_level > 0) {
+      loop = loops[current_level - 1];
+    }
+    auto it = common_index_map_.find(loop);
```

Comment:
A bit lost here. `loop` can be `nullptr`. Is it safe to do lookup with `nullptr`?

Reply:
Why does the loop start with `current_level=0`? Doesn't `level` mean the loop position where all the loop dependency is resolved for this `index`? So, could there be an equivalent index in the loops from 0th to `level-1`-th positions? Isn't it just sufficient to check `loops[level]`?

Reply:
I think nullptr lookup is fine. It is just an integer zero, it will not access the object.
I need something to refer to the top level exprs, which has no corresponding loop. I can either use 0 to refer to top level exprs and use 1 to refer to `loops[0]`, or use -1 to refer to top level exprs and use 0 to refer to `loops[0]`

Reply:
OK, I still don't follow why you need to loops that are outer than the `level` loop. Is it because some subexpression of `index` may be hoisted at those outer loops?

Reply:
> OK, I still don't follow why you need to loops that are outer than the `level` loop. Is it because some subexpression of `index` may be hoisted at those outer loops?

Oh, I think you are right. We don't need it. I originally wrote this thinking that some subexpression of `index` might be hoisted at outer loops. I thought that I would find subexpressions that are both in `index` and `common_index_map_`. But seems that I eventually didn't do that. All I am doing in this function is to find something `sameAs` `index`, I am not further breaking `index` down into pieces(the breakdown of `index` is done in `hoistIndex`). So I think checking `loops[pos]` is sufficient.

Reply:
I updated this PR, did some cleanup with this new understanding. Now `pos == -1` means top level exprs, `pos == 0` means `loops[0]`.